hw/xbox/mcpx_apu.c

```diff
@@ -802,12 +828,104 @@ static const MemoryRegionOps ep_ops = {
     .write = ep_write,
 };
 
-/* TODO: this should be on a thread so it waits on the voice lock */
-static void se_frame(void *opaque)
+static void process_voice(MCPXAPUState *d,
+                          int32_t mixbin[NUM_MIXBINS][NUM_SAMPLES_PER_FRAME],
+                          uint32_t voice)
 {
-    MCPXAPUState *d = opaque;
-    timer_mod(d->se.frame_timer, qemu_clock_get_ms(QEMU_CLOCK_VIRTUAL) + 10);
-    MCPX_DPRINTF("mcpx frame ping\n");
+    /* FIXME: Implement */
+}
+
+#if DUMP_VP_BINS_MASK > 0
+static void dump_vp_bins(MCPXAPUState *d,
+                         int32_t mixbin[NUM_MIXBINS][NUM_SAMPLES_PER_FRAME])
+{
+    for(int i = 0; i < NUM_MIXBINS; i++) {
+        if (d->vp_bin_files[i] == NULL) {
+            continue;
+        }
+        fwrite(&mixbin[i][0], 1, sizeof(mixbin[i]), d->vp_bin_files[i]);
+    }
+}
+#endif
```

Comment:
Should be done in `route_vp_downmix`, because analogous behaviour in GP needs more processing, and will also be necessary for QEMU audio-card anyway. This extra function woul just lead to redundant code